### PR TITLE
Add configurable optimization criterion name

### LINKED ISSUE #94: Configurable optimization criterion name
So far, the base net specifies `loss` as the optimization criterion. It would be nice to be able to specify in in the config (and overwrite the default `loss`).

Implementing this would allow users to pretrain the net using loss `L1` and then finetune the model (in the separate training) to some other loss `L2`

### PR DESCRIPTION
**Pull request by luxusko**
resolves #94

### CODE REVIEW

**Comment by luxusko** [comment]
Many cxflow hooks are defaultly looking for ``loss`` and logs/error in unsuccessful cases are non-comprehensible. Potential cxflow issue?

File: cxflow_tensorflow/model.py [hunk 1]
```diff
@@ -147,16 +148,13 @@ class BaseModel(cx.AbstractModel, metaclass=ABCMeta):  # pylint: disable=too-man
     TRAIN_OP_NAME = 'train_op'
     """Expected train op tensor name prefix."""
 
-    LOSS_NAME = 'loss'
```
Comment (blazekadam) [inline cxflow_tensorflow/model.py]: I would keep this renamed as `DEFAULT_LOSS_NAME` and use it for the default `__init__` parameters, tests etc.

File: cxflow_tensorflow/model.py [hunk 2]
```diff
@@ -41,20 +41,21 @@ class GraphTower:
...
     """
 
-    def __init__(self, id_: int, inputs: List[str], outputs: List[str]):
+    def __init__(self, id_: int, inputs: List[str], outputs: List[str], loss_name: str='loss'):
         """
         Create new GraphTower.
         :param id_: tower (gpu) id, towers with negative ids are placed on /cpu:0
         :param inputs: tower input names
         :param outputs: tower output names
+        :param loss_name: expected loss tensor name
```
Comment (blazekadam) [inline cxflow_tensorflow/model.py]: just a nitpick: remove *expected* for consistency with other docs (you may place it basically everywhere without carrying much info)

File: cxflow_tensorflow/model.py [hunk 3]
```diff
@@ -41,20 +41,21 @@ class GraphTower:
     The sub-graphs must be defined in the order corresponding to the tower ids!
     """
 
-    def __init__(self, id_: int, inputs: List[str], outputs: List[str]):
+    def __init__(self, id_: int, inputs: List[str], outputs: List[str], loss_name: str='loss'):
```
Comment (blazekadam) [inline cxflow_tensorflow/model.py]: Use a constant for the default value as suggested later.

File: cxflow_tensorflow/model.py [hunk 4]
```diff
@@ -147,16 +148,13 @@ class BaseModel(cx.AbstractModel, metaclass=ABCMeta):  # pylint: disable=too-man
...
-    """Expected loss tensor name."""
-
     TRAINING_FLAG_NAME = 'cxf_is_training'
     """Training flag variable name."""
 
     def __init__(self,  # pylint: disable=too-many-arguments
                  dataset: Optional[cx.AbstractDataset], log_dir: str, inputs: List[str], outputs: List[str],
                  session_config: Optional[dict]=None, n_gpus: int=0, restore_from: Optional[str]=None,
-                 restore_model_name: Optional[str]=None, optimizer=None, freeze=False, **kwargs):
+                 restore_model_name: Optional[str]=None, optimizer=None, freeze=False, loss_name: str='loss', **kwargs):
```
Comment (blazekadam) [inline cxflow_tensorflow/model.py]: ditto, constant default

File: cxflow_tensorflow/model.py [hunk 5]
```diff
@@ -182,18 +180,20 @@ def __init__(self,  # pylint: disable=too-many-arguments
         :param restore_model_name: model name to be restored (e.g. ``model.ckpt``)
         :param optimizer: TF optimizer configuration dict
         :param freeze: freeze the graph after each save
+        :param loss_name: expected loss tensor name
```
Comment (blazekadam) [inline cxflow_tensorflow/model.py]: ditto *expected*

File: cxflow_tensorflow/tests/model_test.py [hunk 6]
```diff
@@ -16,13 +16,13 @@
```
Comment (blazekadam) [inline cxflow_tensorflow/tests/model_test.py]: The tests have unnecessary large diff. Just leave it to the loss name to the default value and add a single test case covering the custom loss name.

File: cxflow_tensorflow/model.py [hunk 7]
```diff
@@ -14,6 +14,10 @@
 from .utils import create_optimizer
 
 
+DEFAULT_LOSS_NAME = 'loss'
+"""Default loss tensor name."""
```
Comment (luxusko) [inline cxflow_tensorflow/model.py]: This might be wierd but  ``GraphTower.__init__`` int woudln't take default value ``BaseModel.LOSS_NAME`` (`BaseModel not defined`). Possible other solution is swap order of `GraphTower` and `BaseModel` definitions which I find very ugly. In general constants used not only in one particular class shouldn't be in one particular class.

Comment (blazekadam) [reply cxflow_tensorflow/model.py]: I have no issue with it. Thanks.
